### PR TITLE
kloud: new stack API

### PR DESCRIPTION
**Pull request by rjeczalik**
~~deps: https://github.com/koding/koding/pull/9004~~

### CODE REVIEW

**Comment by rjeczalik** [comment]
@cihangir @ppknap Take your time reviewing this, I'm going to update description with summary of changes a bit later.

**Comment by rjeczalik** [comment]
@cihangir @ppknap I've pushed documentation of this change [here](https://github.com/koding/kloud-provider-example). The purpose of this PR is to reduce boilerplate needed to create a new provider, like [in this diff](https://github.com/koding/kloud-provider-example/commit/e9ddd78415590be43cf991bdf41d198c4c365997).

ptal

File: go/src/koding/kites/kloud/stack/provider/stacker.go [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+	"golang.org/x/net/context"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// TODO(rjeczalik): remove context.Context from the Provider / Stack / Machine API,
+// which is a leftover from the old API - it was errornously desgined to
+// store global-scoped values.
+
+// Stacker is responsible for augementing Provider values so they implement
```
Comment (cihangir) [inline go/src/koding/kites/kloud/stack/provider/stacker.go]: typo augementing

File: go/src/koding/kites/kloud/stack/provider/template.go [hunk 2]
```diff
@@ -227,23 +228,37 @@ func (t *Template) FillVariables(prefix string) error {
...
+func (t *Template) inject(prefix string, meta interface{}) {
+	t.log.Debug("injecting: %# v", meta)
+
 	for k, v := range t.b.New(prefix).Build(meta) {
 		// Ignore custom variables prefixed with __ from injecting.
 		// The ignored variables may contain interpolations which are
 		// not meant to be sent to Terraform.
 		if strings.HasPrefix(k, "custom___") {
-			t.log.Debug("Not injecting variable: %s=%s", k, v)
+			t.log.Debug("not injecting variable: %s=%s", k, v)
```
Comment (cihangir) [inline go/src/koding/kites/kloud/stack/provider/template.go]: we were trying to follow "errors are values/variables so they should start with small char, logs are sentences, should start with sentence case" rule.

File: go/src/koding/kites/kloud/provider/aws/aws.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+func init() {
+	provider.Register(p)
+}
+
+func newMachine(bm *provider.BaseMachine) (provider.Machine, error) {
+	m := &Machine{BaseMachine: bm}
+	cred := m.Cred()
+
+	opts := &amazon.ClientOptions{
+		Credentials: credentials.NewStaticCredentials(cred.AccessKey, cred.SecretKey, ""),
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/aws/aws.go]: To simplify a bit: `Credentials: cred.Credentials(),`

**Review by cihangir — APPROVED** [review]

File: go/src/koding/kites/kloud/provider/aws/aws_test.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package aws_test
+
+import (
+	"koding/kites/kloud/provider/aws"
+	"koding/kites/kloud/stack"
+)
+
+var (
+	_ stack.Machiner = (*aws.Machine)(nil)
+	_ stack.Stacker  = (*aws.Stack)(nil)
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/aws/aws_test.go]: Why are these checks not in `machine.go`/`stack.go` files? This is inconsistent with`stack.Validator` interface checks in `aws.go`.

Comment (ppknap) [inline go/src/koding/kites/kloud/provider/aws/aws_test.go]: Should these checks be here?

File: go/src/koding/kites/kloud/provider/aws/stack.go [hunk 5]
```diff
@@ -1,218 +1,268 @@
...
+func (s *Stack) VerifyCredential(c *stack.Credential) error {
+	cred := c.Credential.(*Cred)
+
+	if err := cred.Valid(); err != nil {
+		return err
 	}
-	if meta.CidrBlock == "" {
-		return errors.New("CIDR block is empty or missing")
+
+	_, err := amazon.NewClient(cred.Options())
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/aws/stack.go]: What if AWS credentials have custom policy that eg. doesn't allow to create EC2 instances? Can't we check it here?

File: go/src/koding/kites/kloud/provider/aws/stack.go [hunk 6]
```diff
@@ -1,218 +1,268 @@
...
+	return err
+}
+
+func (s *Stack) BootstrapTemplates(c *stack.Credential) ([]*stack.Template, error) {
+	cred := c.Credential.(*Cred)
+
+	opts := cred.Options()
+	opts.Log = s.Log.New("amazon")
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/aws/stack.go]: Related to first comment

File: go/src/koding/kites/kloud/provider/aws/stack.go (hunk 5 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/aws/stack.go]: There's a bigger story about that - for now we require AWS creds to have full EC2 support + IAM::GetUser permission (to read amazon account ID). That's how we document requirement for AWS credential. I wouldn't want only to make EC2 checks here, I'd want to drop this requirement, so it's possible to run AWS stacks with less privileged user. This however requires a number of patches. Created https://github.com/koding/koding/issues/9137 to track this.

File: go/src/koding/kites/kloud/provider/vagrant/stack.go [hunk 7]
```diff
@@ -1,82 +1,275 @@
...
+			Groups:          []string{"sudo"},
+			Hostname:        s.Req.Username, // no typo here. hostname = username
+			KiteKey:         kiteKey,
+			LatestKlientURL: klientURL,
+			TunnelName:      tunnel.Name,
+			TunnelKiteURL:   tunnel.KiteURL,
+			KontrolURL:      kontrolURL,
+		}
+
+		// pass the values as a JSON encoded as bae64. Our script will decode
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/vagrant/stack.go]: `base64`

File: go/src/koding/kites/kloud/stack/provider/builder.go [hunk 8]
```diff
@@ -483,27 +486,31 @@ func (b *Builder) BuildCredentials(method, username, groupname string, identifie
...
-		b.Credentials = append(b.Credentials, cred)
+	if err := b.FetchCredentials(username, creds...); err != nil {
+		// TODO(rjeczalik): add *NotFoundError support to CredStore
+		return ResError(err, "jCredentialData")
 	}
 
-	b.Log.Debug("Built credentials: %+v", b.Credentials)
+	b.Credentials = append(b.Credentials, creds...)
+
+	b.Log.Debug("Built credentials: %# v", b.Credentials)
```
Comment (ppknap) [inline go/src/koding/kites/kloud/stack/provider/builder.go]: Although this works, I'd rather you removed the space from `%# v`.

**Review by ppknap — COMMENTED** [review]
@rjeczalik lgtm

**Pull request merged**